extensions/wikia/ImageReview/ImageReviewHelper.class.php

```diff
@@ -203,221 +235,283 @@ public function refetchImageListByTimestamp( $timestamp ) {
-							'wikiId' => $row->wiki_id,
-							'pageId' => $row->page_id,
-							'state' => $row->state,
-							'src' => $aImageInfo['src'],
-							'url' => $aImageInfo['page'],
-							'priority' => $row->priority,
-							'flags' => $row->flags,
-							'isthumb' => $isThumb,
-							'wiki_url' => isset( $wikiRow->city_url ) ? $wikiRow->city_url : '',
-						);
-					}
-				} else {
-					$aDeleteFromQueueList[] = [
-						'wiki_id' => $row->wiki_id,
-						'page_id' => $row->page_id,
-						'reason' => $sReason,
-					];
-					continue;
+		// Move icons to different state
+		$this->moveImagesToIcoStateFromWhere( $iconsWhere );
+
+		// Return valid images list
+		WikiaLogger::instance()->info( 'ImageReviewLog', [
+			'method' => __METHOD__,
+			'message' => 'Fetched ' . count( $imageList ) . ' new images',
+		] );
+
+		return $imageList;
+	}
+
+	/**
+	 * Return an iterator that keeps fetching images from the review table that are in state $state.
+	 * It is the callers responsibility to either update the value of the state column in the DB
+	 * such that this iterator will eventually run out of matching images, or end the loop
+	 * explicitly after a set number of rows.
+	 *
+	 * @param $order
+	 * @param $state
+	 * @return Closure
+	 */
+	private function getLatestImageIter( $order, $state ) {
+		$dbh = $this->getDatawareDB( DB_SLAVE );
+		$helper = $this->getDatabaseHelper();
+		$orderBy = $this->getOrder( $order );
+		$limit = self::LIMIT_IMAGES_FROM_DB;
+
+		return function () use ( $dbh, $helper, $orderBy, $limit, $state ) {
+			static $rows;
+
+			if ( empty( $rows ) ) {
+				$results = $helper->selectImagesForList( $orderBy, $limit, $state );
+				if ( $results->numRows() == 0 ) {
+					$dbh->freeResult( $results );
+
+					return null;
+				}
+
+				while ( $row = $dbh->fetchObject( $results ) ) {
+					$rows[] = $row;
 				}
-			} else {
-				$unusedImages[] = $record;
+				$dbh->freeResult( $results );
 			}
+
+			return array_pop( $rows );
+		};
+	}
+
+	private function assignImageToUser( $reviewStart, $state, stdClass $imageRecord ) {
+		// Determine what our next state should be
+		$targetState = ImageReviewStatuses::STATE_IN_REVIEW;
+
+		if ( $state == ImageReviewStatuses::STATE_QUESTIONABLE ) {
+			$targetState = ImageReviewStatuses::STATE_QUESTIONABLE_IN_REVIEW;
+		} elseif ( $state == ImageReviewStatuses::STATE_REJECTED ) {
+			$targetState = ImageReviewStatuses::STATE_REJECTED_IN_REVIEW;
 		}
 
-		/**
-		 * Invalid images
-		 */
-		if ( !empty( $aDeleteFromQueueList ) ) {
-			$this->createDeleteFromQueueTask( $aDeleteFromQueueList );
+		$dbw = $this->getDatawareDB( DB_MASTER );
+		$query = ( new WikiaSQL() )
+			->UPDATE( 'image_review' )
+				->SET( 'reviewer_id', $this->user_id )
+				->SET( 'review_start', $reviewStart )
+				->SET( 'state', $targetState )
+			->WHERE( 'reviewer_id' )->IS_NULL()
+				->AND_( 'wiki_id' )->EQUAL_TO( $imageRecord->wiki_id )
+				->AND_( 'page_id' )->EQUAL_TO( $imageRecord->page_id );
```

Comment:
By setting a reviewer_id and checking if its null in the where at the same time we ensure don't steal an image being reviewed by someone else.